# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Libraries
import numpy as np
import pandas as pd
from sklearn.datasets import load_diabetes
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import scale
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score
import statsmodels.api as sm

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# your code here
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# your code here
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# your code here
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

#### Data description
1. & 3.- There are 6 attributes in the data dictionary: ['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename']. Basically 'data' contains the features of the model (characteristics and health indicators of a sample of 442 people), the 'target' being a quantitative evaluation of diabetes progression (one year after the health analysis), and the remaining keys having a description of the dataset, feature names and path to the 'csv' containing data.
2. - Basically what we want to learn through the data using machine learning is how the different characteristics and health indicators from people - that is: `diabetes['data']` - influence the evolution of the quantitative measurement of diabetes progression - that is `diabetes['target']` -.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# your code here
print(f"Data contains {diabetes.data.shape[1]} model features for {diabetes.data.shape[0]} \
records in a numpy array of type: {diabetes.data.dtype}.")
print(f"Target is a vector (numpy array) containing {diabetes.target.shape[0]} records \
of type: {diabetes.target.dtype}.")

Data contains 10 model features for 442 records in a numpy array of type: float64.
Target is a vector (numpy array) containing 442 records of type: float64.


# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:


$y = \beta_0 + \beta_1·X_1 + \beta_2·X_2 + ... + \beta_n·X_n + \epsilon$


where:

- $X_1-X_n$: data 
- $\beta_0$: intercept 
- $\beta_1-\beta_n$: coefficients 
- $\epsilon$: error (cannot explained by model)
- $y$: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

Imported in first cell as: `from sklearn.linear_model import LinearRegression`


#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [6]:
# your code here
# give data is already mean centered, then intercept is set to False
diabetes_model = LinearRegression(fit_intercept=False)

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

Given that the records are already random (not ordered by age or sex), we can proceed using the last 20 records for testing (without need of shuffling data, that is `shuffle=False`). Actually the features are scaled but not the target. Having set intercept to zero, then I must scaled the target using the mean-centered approach.

In [7]:
# your code here
X = diabetes.data
y = diabetes.target
diabetes_data_train, diabetes_data_test, diabetes_target_train, diabetes_target_test = \
train_test_split(X,y,test_size=20,shuffle=False)
# scale target using mean-centered and std scaling
y_train_scale = scale(diabetes_target_train)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [8]:
# your code here
lin_reg = diabetes_model.fit(diabetes_data_train, y_train_scale)
print(f"The coefficients of our multivariable linear regression are: {lin_reg.coef_}")
print(f"The intercept is zero given our input is all mean-centered.")

The coefficients of our multivariable linear regression are: [ 1.77122757e-03 -3.07990853e+00  6.61747190e+00  4.24806059e+00
 -1.05648755e+01  6.40019849e+00  1.33919480e+00  2.39288478e+00
  9.64389523e+00  9.89893379e-01]
The intercept is zero given our input is all mean-centered.


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [9]:
# your code here
print('The predictions of the model are below:')
print(lin_reg.predict(diabetes_data_test))

The predictions of the model are below:
[ 0.58181843  0.03461461  0.26094966 -0.53459643  0.15645203 -0.28121945
  1.37909092 -0.67765437 -0.46279422 -0.36904634  0.85086795 -1.18676962
 -0.26587629 -0.42058631 -1.29920974  0.53500167 -0.65031694 -0.37865548
  0.75539064 -1.29782859]


#### Print your `diabetes_target_test` and compare with the prediction. 

In [10]:
# your code here
# scale target test to compare
mean_target_train = diabetes_target_train.mean()
std_target_train  = diabetes_target_train.std()
print('Below we observe the target test values (mean-centered scaled):')
target_test_scaled = (diabetes_target_test-mean_target_train)/std_target_train
print(target_test_scaled)

Below we observe the target test values (mean-centered scaled):
[ 1.0325496  -0.80856987 -0.54925727 -0.01766644 -0.4325666  -1.11974499
  2.03090311 -0.76967298  0.38426809 -1.13271062  0.25461179 -1.05491684
 -1.35312634 -1.15864188 -1.36609197  0.31943994 -0.64001668 -0.27697904
  0.86399641 -1.2494013 ]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

The *target test* and the *predictions* significantly differ one another. This means the multivariable linear regression we have performed is not a good model for predictions. We could further assess these difference by computing the linear regression score ($R^2$) and also looking at the absolute relative differences between the 20 components of *predictions* array against those from *target test* array.

In [11]:
# array of absolute relative differences in %
np.set_printoptions(precision=2,suppress=True)
r2_lin_reg = diabetes_model.score(diabetes_data_train, y_train_scale)
print(f"The R^2 coefficient for the multivariable regression is: {r2_lin_reg.round(2)}")
rel_def = (abs(lin_reg.predict(diabetes_data_test)-target_test_scaled)/target_test_scaled*100).round(2)
print(f"The array of absolute relative diferences [%] is: {rel_def}")

The R^2 coefficient for the multivariable regression is: 0.51
The array of absolute relative diferences [%] is: [   43.65  -104.28  -147.51 -2926.06  -136.17   -74.89    32.09   -11.96
   220.44   -67.42   234.18   -12.5    -80.35   -63.7     -4.9     67.48
    -1.61   -36.71    12.57    -3.88]


# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://online.stat.psu.edu/stat501/lesson/6/6.4) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    $H_0 : \beta_1 = \beta_2 = ... = \beta_{10} = 0$

    $H_a$ : At least one $\beta_j \ne 0$ (for j = 1, 2, ..., 10)
    
    
* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

<div style="text-align:center"><img src="../images/statsmodels.png" width=600/></div>

In [12]:
# your code here
# Fit and summarize OLS model
mod = sm.OLS(y_train_scale,diabetes_data_train)
res = mod.fit()
print(res.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.512
Model:                            OLS   Adj. R-squared (uncentered):              0.500
Method:                 Least Squares   F-statistic:                              43.25
Date:                Wed, 20 May 2020   Prob (F-statistic):                    3.36e-58
Time:                        04:33:37   Log-Likelihood:                         -447.34
No. Observations:                 422   AIC:                                      914.7
Df Residuals:                     412   BIC:                                      955.1
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

1. - The F-score of the model is 43.25 and the null hypothesis is not rejected.
2. - Yes, there are many coefficients whose confidence intervals contains zero ($x_1$, $x_5$, $x_6$, $x_7$, $x_8$, $x_{10}$).
3. - I would basically eliminate 6 features from the data set (those corresponding to columns: 1, 5 to 8 - both inclusive - and 10).

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [13]:
# your code here
auto = pd.read_csv('../data/auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [14]:
# your code here
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [15]:
# your code here
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horse_power   392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   car_name      398 non-null    object 
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


What is the newest model year and the oldest model year?

In [16]:
# your code here
print(f"The newest year model in dataset is {auto.model_year.max()}.")
print(f"The oldest year model in dataset is {auto.model_year.min()}.")

The newest year model in dataset is 82.
The oldest year model in dataset is 70.


Check the dataset for missing values and remove all rows containing at least one missing value.

In [17]:
# your code here
print('Check for NaNs:')
print(auto.isna().sum())
auto.dropna(inplace=True)
print('Check after dropping NaNs:')
print(auto.isna().sum())

Check for NaNs:
mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64
Check after dropping NaNs:
mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64


Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [18]:
# your code here
print(auto.cylinders.value_counts(sort=False))
print(f"There are cars with these cylinders: {auto.cylinders.value_counts(sort=False).index.values}")

3      4
4    199
5      3
6     83
8    103
Name: cylinders, dtype: int64
There are cars with these cylinders: [3 4 5 6 8]


We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [19]:
# your code here
X = auto.drop(columns=['mpg','car_name'])
y = auto.mpg
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [20]:
# your code here
auto_model  = LinearRegression()
mpg_lin_reg = auto_model.fit(X_train, y_train)

In [21]:
# create same model but using mean-centered & scaled data to compare results (expected: the very same for linear regression)
auto_model_centered = LinearRegression(fit_intercept=False)
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
y_train_std = scale(y_train)
mpg_lin_reg_std = auto_model_centered.fit(X_train_std, y_train_std)


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). Its formula is:

<div style="text-align:center"><img src="../images/r-squared.png" width=300/></div>

where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [22]:
# your code here (the indications are indeed confusing because SSE is not r^2)
# compute r^2 by myself
def r2_score_by_beto(y_true, y_predict):
    # Compute SSE a.k.a the sum of the squared residuals (assuming np array)
    ss_res = ((y_true - y_predict)**2).sum()
    # Compute the total sum of squares
    ss_tot = ((y_true - y_true.mean())**2).sum()
    return 1 - ss_res/ss_tot
y_pred = mpg_lin_reg.predict(X_train)
print(f"The r2 score computed for the training data results {round(r2_score_by_beto(y_train,y_pred),2)}")

The r2 score computed for the training data results 0.81


#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [23]:
# your code here
y_test_pred = mpg_lin_reg.predict(X_test)
print(f"The r2 score computed for the test data results {round(r2_score_by_beto(y_test,y_test_pred),2)}")

The r2 score computed for the test data results 0.79


#### Explaining the results

Please, compare the results for the training and test sets and comment. What can you say?

We can conclude that the r2 score behaves similarly between train and test data, therefore the model is not overfitted. However, it is still arguable if the model accuracy could be made higher than 0.79 for the test data.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [24]:
# your code here
X_train09, X_test09, y_train09, y_test09 = train_test_split(X,y,test_size=0.1,random_state=42)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [25]:
# your code here
auto_model09  = LinearRegression()
mpg_lin_reg_09 = auto_model09.fit(X_train09, y_train09)

Compute the predicted values and r squared score for our new model and new sample data.

In [26]:
# your code here
y_pred09 = mpg_lin_reg_09.predict(X_train09)
print(f"The r2 score computed for the training data 90% results {round(r2_score_by_beto(y_train09,y_pred09),2)}")

The r2 score computed for the training data 90% results 0.8


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [27]:
# your code here
y_test_pred09 = mpg_lin_reg_09.predict(X_test09)
print(f"The r2 score computed for the test data 10% results {round(r2_score_by_beto(y_test09,y_test_pred09),2)}")

The r2 score computed for the test data 10% results 0.85


The r2 score remains more or less the same for the training data but improves for the test data, given that the new model is better trained.

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE.

In [28]:
# Libraries
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`.

In [29]:
# your code here

Fit the model and print the ranking.

In [30]:
# your code here

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [31]:
# your code here

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [32]:
# your code here

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.